### ISSUE TITLE
Wasm bindings

### ISSUE THREAD

**Issue by arcanis** [DioxusLabs/taffy#241]
## What problem does this solve or what need does it fill?

It seems that the wasm bindings don't exist anymore, so it's difficult to use this library from JS-land. I made a [JS terminal library](https://github.com/manaflair/mylittledom) many years ago using Yoga and I'd like to rebuild it, evaluating new tech options.

## What solution would you like?

Ideally a basic wasm API would be great. If not possible, it'd be useful to at least annotate FlexboxLayout & its enums with the wasm-bindgen attributes (cf second part of this answer here: https://stackoverflow.com/a/65336309), as there's a lot of them and duplicating them would be very painful.

## What alternative(s) have you considered?

Keep using Yoga, but I'm interested to follow a bit the development here.

**Comment by nicoburns** [comment]
I've been looking into this, and it looks like wasm support might not be too difficult:

- The predecessor to this library `stretch` had wasm bindings ([source here](https://github.com/vislyhq/stretch/tree/master/bindings/js) and [docs here](https://vislyhq.github.io/stretch/docs/javascript))
- The setup seems to be a [vanilla wasm-pack setup](https://rustwasm.github.io/wasm-pack/book/tutorials/npm-browser-packages/index.html)
- The [meat of the bindings](https://github.com/vislyhq/stretch/blob/master/bindings/js/src/lib.rs) aren't all that long (650 lines with quite a bit of boilerplate). And it looks like some of the boilerplate could be cut down on by adding `#[wasm_bindgen]` directly to taffy types (behind a feature flag) rather than duplicating them (if we are willing to bring in another dependency then `num-derive` and/or `serde-wasm-bindgen` could further cut down on boilerplate)

**Comment by alice-i-cecile** [comment]
Awesome! I'd be happy to review that.

**Comment by theetherGit** [comment]
> I've been looking into this, and it looks like wasm support might not be too difficult:
> 
> * The predecessor to this library `stretch` had wasm bindings ([source here](https://github.com/vislyhq/stretch/tree/master/bindings/js) and [docs here](https://vislyhq.github.io/stretch/docs/javascript))
> * The setup seems to be a [vanilla wasm-pack setup](https://rustwasm.github.io/wasm-pack/book/tutorials/npm-browser-packages/index.html)
> * The [meat of the bindings](https://github.com/vislyhq/stretch/blob/master/bindings/js/src/lib.rs) aren't all that long (650 lines with quite a bit of boilerplate). And it looks like some of the boilerplate could be cut down on by adding `#[wasm_bindgen]` directly to taffy types (behind a feature flag) rather than duplicating them (if we are willing to bring in another dependency then `num-derive` and/or `serde-wasm-bindgen` could further cut down on boilerplate)

I found [napi-rs](https://github.com/napi-rs/napi-rs) used by resvg-js and it's pretty good binding with node too.

**Comment by nicoburns** [comment]
> I found [napi-rs](https://github.com/napi-rs/napi-rs) used by resvg-js and it's pretty good binding with node too.

napi-rs does indeed look like a good option for node bindings (which yoga also has), however I don't think that would give us wasm support. So I would consider that a separate feature request.

**Comment by nicoburns** [comment]
@xavierzwirtz Yeah, when I looked into this recently, it looked like C and WASM would be both be quite straightforward. But stretch's Swift and especially it's Kotlin bindings looked quite a bit more involved. Perhaps we ought to create a new meta issue for language bindings. It would be good to determine which languages have the most interest, and perhaps come up with a strategy that can allow us to re-add the bindings while keeping the maintenance burden manageable.

I suspect focussing on C and WASM first might be a good strategy. WASM effectively gives us Node support too (albeit with a performance penalty), and I suspect a good C API would make it easier for other people to write their own bindings for other languages, which generally support C FFI.
